test/e2e/cleanup/submariner.go

```diff
@@ -0,0 +1,330 @@
+			names.AppendUninstall(names.NetworkPluginSyncerComponent), metav1.GetOptions{})
+		Expect(apierrors.IsNotFound(err)).To(BeTrue(), fmt.Sprintf("Unexpected Deployment %q found",
+			names.AppendUninstall(names.NetworkPluginSyncerComponent)))
+
+		assertNoClusterResources(brokerRestConfig, submariner.Spec.BrokerK8sRemoteNamespace)
+		assertNoEndpointResources(brokerRestConfig, submariner.Spec.BrokerK8sRemoteNamespace)
+		assertNoGlobalnetResources()
+	})
+}
+
+func assertNoGlobalnetResources() {
+	client, err := submarinerclientset.NewForConfig(framework.RestConfigs[framework.ClusterA])
+	Expect(err).To(Succeed())
+
+	list, err := client.SubmarinerV1().ClusterGlobalEgressIPs(metav1.NamespaceNone).List(context.TODO(), metav1.ListOptions{})
+	Expect(err).To(Succeed())
+	Expect(list.Items).To(BeEmpty())
+}
+
+func assertNoEndpointResources(brokerRestConfig *rest.Config, brokerNS string) {
+	clusterID := framework.TestContext.ClusterIDs[framework.ClusterA]
+
+	By(fmt.Sprintf("Verifying no Endpoint resources for %q", clusterID))
+
+	assertNoEndpoints := func(restConfig *rest.Config, namespace, msgFormat string) {
+		client, err := submarinerclientset.NewForConfig(restConfig)
+		Expect(err).To(Succeed())
+
+		endpoints, err := client.SubmarinerV1().Endpoints(namespace).List(context.TODO(), metav1.ListOptions{})
+		Expect(err).To(Succeed())
+
+		for i := range endpoints.Items {
+			Expect(endpoints.Items[i].Spec.ClusterID).ToNot(Equal(clusterID), fmt.Sprintf(msgFormat, endpoints.Items[i].Name))
+		}
+	}
+
+	assertNoEndpoints(brokerRestConfig, brokerNS, "Unexpected Endpoint %q found on the broker")
+	assertNoEndpoints(framework.RestConfigs[framework.ClusterA], framework.TestContext.SubmarinerNamespace,
+		"Unexpected local Endpoint %q found")
+}
+
+func assertNoClusterResources(brokerRestConfig *rest.Config, brokerNS string) {
+	clusterID := framework.TestContext.ClusterIDs[framework.ClusterA]
+
+	By(fmt.Sprintf("Verifying no Cluster resources for %q", clusterID))
+
+	assertNoCluster := func(restConfig *rest.Config, namespace, msgFormat string) {
+		client, err := submarinerclientset.NewForConfig(restConfig)
+		Expect(err).To(Succeed())
+
+		_, err = client.SubmarinerV1().Clusters(namespace).Get(context.TODO(), clusterID, metav1.GetOptions{})
+		Expect(apierrors.IsNotFound(err)).To(BeTrue(), fmt.Sprintf(msgFormat, clusterID))
+	}
+
+	assertNoCluster(brokerRestConfig, brokerNS, "Unexpected Cluster %q found on the broker")
+	assertNoCluster(framework.RestConfigs[framework.ClusterA], framework.TestContext.SubmarinerNamespace,
+		"Unexpected local Cluster %q found")
+}
+
+type podInfo struct {
+	status  corev1.PodStatus
+	log     string
+	prevLog string
+}
+
+type podMonitor struct {
+	sync.Mutex
+	name string
+	pods map[string]*podInfo
+}
+
+func startPodMonitor(name string, stopCh <-chan struct{}) *podMonitor {
+	m := &podMonitor{
+		name: name,
+		pods: map[string]*podInfo{},
+	}
+
+	w, err := watcher.New(&watcher.Config{
```

Comment:
This pod monitor watcher pattern is interesting, might be a good topic for a code teaching session.